### PR TITLE
Gpescia/continuos operator

### PR DESCRIPTION
**Pull request by gpescia**
I started to change the current `AbstractOperator` class to a more general one suitable for discrete and continuous operators, and added `DiscreteOperator` class for all the operators that have already been implemented. Tell me what you think.

Question:
 I left methods in the AbstractOperator class that rely on `._lazy` which now takes DiscreteOperator as a parent. I think there should be a second ._lazy file for Continuous operators (or dispatching) or the methods relying on ._lazy are transferred to DiscreteOperator. What do you think?

### CODE REVIEW

File: netket/operator/_abstract_operator.py [hunk 1]
```diff
@@ -123,224 +104,3 @@ def conjugate(self, *, concrete=False) -> "AbstractOperator":
...
-            _get_conn_flattened_closure not implemented for this operator type.
-            You were probably trying to use an operator with a sampler.
-            Please report this bug.
-
-            numba4jax won't work.
-            """
-        )
-
-    def __repr__(self):
-        return f"{type(self).__name__}(hilbert={self.hilbert})"
```
Comment (PhilipVinc) [inline netket/operator/_abstract_operator.py:311]: you can leave this as a fallback

File: netket/operator/_abstract_operator.py [hunk 2]
```diff
@@ -123,224 +104,3 @@ def conjugate(self, *, concrete=False) -> "AbstractOperator":
...
-    def apply(self, v: np.ndarray) -> np.ndarray:
-        op = self.to_linear_operator()
-        return op.dot(v)
-
-    def __call__(self, v: np.ndarray) -> np.ndarray:
-        return self.apply(v)
-
-    def conj(self, *, concrete=False) -> "AbstractOperator":
-        return self.conjugate(concrete=False)
-
```
Comment (PhilipVinc) [inline netket/operator/_abstract_operator.py:295]: also this one

File: netket/operator/_discrete_operator.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            _get_conn_flattened_closure not implemented for this operator type.
+            You were probably trying to use an operator with a sampler.
+            Please report this bug.
+
+            numba4jax won't work.
+            """
+        )
+
+    def __repr__(self):
+        return f"{type(self).__name__}(hilbert={self.hilbert})"
```
Comment (PhilipVinc) [inline netket/operator/_discrete_operator.py:206]: remove and move to base

File: netket/operator/_discrete_operator.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+    def apply(self, v: np.ndarray) -> np.ndarray:
+        op = self.to_linear_operator()
+        return op.dot(v)
+
+    def __call__(self, v: np.ndarray) -> np.ndarray:
+        return self.apply(v)
+
+    def conj(self, *, concrete=False) -> "DiscreteOperator":
+        return self.conjugate(concrete=False)
```
Comment (PhilipVinc) [inline netket/operator/_discrete_operator.py:189]: remove and move to base

File: netket/operator/_abstract_operator.py [hunk 5]
```diff
@@ -123,224 +104,3 @@ def conjugate(self, *, concrete=False) -> "AbstractOperator":
...
-        low-dimensional Hilbert spaces or sufficiently sparse operators.
-
-        This method requires an indexable Hilbert space.
-
-        Returns:
-            The dense matrix representation of the operator as a Numpy array.
-        """
-        return self.to_sparse().todense().A
-
-    def apply(self, v: np.ndarray) -> np.ndarray:
```
Comment (PhilipVinc) [inline netket/operator/_abstract_operator.py:286]: also this one can be left here, with a definition of what it does and a NotImplementedError

File: netket/operator/_abstract_operator.py [hunk 6]
```diff
@@ -123,224 +104,3 @@ def conjugate(self, *, concrete=False) -> "AbstractOperator":
...
-            Please report this bug.
-
-            numba4jax won't work.
-            """
-        )
-
-    def __repr__(self):
-        return f"{type(self).__name__}(hilbert={self.hilbert})"
-
-    def __matmul__(self, other):
```
Comment (PhilipVinc) [inline netket/operator/_abstract_operator.py:313]: everything below here should stay in here.
i will rewrite that stuff with dispatch at some point, but for now let's leave it here

**Comment by PhilipVinc** [comment]
Everything in `Lazy` should stay the same. 
Those are just lazy wrappers to signal that an operator is transposed/conjugatetransposed/squared to dispatch to more efficient implementations.
They assume nothing of the underlying structure.

**Comment by gcarleo** [comment]
Just to add something here on the requirements: I think that a general view on operators could be that they allow to

1.  for an arbitrary basis element |X> compute the action O|X> symbolically (in a lazy way). For discrete operators the concrete version will return the current implementation of get_conn, for continous operators, this should return a differential operator 

2. This information must be usable then to compute the matrix elements (in log space)  log(<X|O|Psi>), for some variational state Psi 

The thing that is slightly tricky here is how to fulfill requirement 2 knowing that the the model variational state yields log<X|Psi>

**Comment by gcarleo** [comment]
@gpescia do we have a nice (or just implementable) expression for log(<X|H|Psi>) in terms of linear operators acting on log(<X|Psi>)  ?

**Comment by gpescia** [comment]
> @gpescia do we have a nice (or just implementable) expression for log(<X|H|Psi>) in terms of linear operators acting on log(<X|Psi>) ?

Yes there is. If H is a differential operator of e.g. order 2 and the wave-function is given by psi(x) = exp(l(x)) then the expression for log(<X|H|Psi>) is given by the complete Bell polynomial in two variables (B(x1,x2) = x1^2 + x_2, where the subscript denotes the order of the differential,). (https://en.wikipedia.org/wiki/Bell_polynomials : Examples). These polynomials are implemented in SymPy. 
I also found this paper where they introduce "Taylormode autodiff" for higher order differentials, which also uses these polynomials.

I think the implementation that is used in netket depends on what orders of derivatives are actually needed to be implemented. I mean is there a Hamiltonian that needs for example the 5th derivative of the wave-function? If only first and second order are needed, there is no need to make things complicated... 

what do you think?

**Comment by gcarleo** [comment]
great! yes, I would implement this only for the kinetic term part for the moment, I am not aware of many physical hamiltonians with more than second derivatives anyways...

**Comment by gcarleo** [comment]
in any case, I think that it makes sense to define the interface of a general linear operator A as something that is able to compute log(<X|A|Psi>) for an arbitrary element of the basis X and for a (variational) state Psi, this generalizes well to both continuum and lattice I think...

**Comment by PhilipVinc** [comment]
Yes, though i think that it makes more sense to define a lazy `applied operator` O|psi>` rather than `<x|O` that you proposed above because that will play better with batching of `<x|`...

Regardless, i think this is somewhat orthogonal to the aim of this PR, which should just get a function` expecT_and_grad` working. 
A nicer interface for operators is something that would be worthwile doing but i'd leave that for later, once we have a clearer idea of how everything plays out in the continuum.

It would also probably break (or be worth doing by breaking) parts of the interface...

**Comment by gcarleo** [comment]
Yes, the reason I was proposing <X|O is because I wanted to keep the interface of the operator possibly clean from the autograd details, but I think that is not possible anyways... unless one complicates things 

So in conclusion, I would suggest: 

1. Skip for the moment the task of implementing O|X> or O|Psi> directly for general operators, since this needs more thoughts about the lazy implementation in the continuum etc... 
2. Define the general interface requiring the ability to compute log(<X|O|Psi>), where Psi is passed as a variational state (in principle this should be more generally a quantum state, which is the base class we don't have yet for the variational state, but it's fine for now...) 
3. For lattice operators we can keep everything as it is, essentially, we just need to define a method to compute log(<X|O|Psi>) using what we already have

**Comment by PhilipVinc** [comment]
>Yes, the reason I was proposing <X|O is because I wanted to keep the interface of the operator possibly clean from the autograd details, but I think that is not possible anyways... unless one complicates things

In some sense, we could represent `<x|O` as a (lazy) sparse matrix, and then apply it to |psi>...

Sometimes i think that the cleanest implementation would be to go along the line of jVMC and make the operators pytrees that are jittable (with some smarts, using sparse matrices as output i think). 
This would clean up implementations because we could essentially jit everything, though compilation times will probably skyrocket....

Anyhow. Let's not think about this...
-- 

why do we need a method to compute `log(X O PSi)`? i agree that it would be nice, but it's not really needed (or is it?)
If it's not essential to this PR i'd leave it for later too...

**Comment by gcarleo** [comment]
so the problem I see here is that expect_and_grad works using dispatch, what I am proposing is easier (I think) because I am saying that if you require all operators to be able to compute log(<X|O|Psi>) then the expression for expect and grad does not need to be dispatched, it's just the same for all cases

**Comment by gpescia** [comment]
> why do we need a method to compute log(X O PSi)? i agree that it would be nice, but it's not really needed (or is it?)
If it's not essential to this PR i'd leave it for later too...
you need the elements to compute expect_and_grad...
At the moment I just implemented log(<x|d^2/dx^2 + V(x)|psi>) to compute energy expectation and gradient. This already works, but it does not take arbitrary differential operators.

**Comment by gcarleo** [comment]
> At the moment I just implemented log(<x|d^2/dx^2 + V(x)|psi>) to compute energy expectation and gradient. This already works, but it does not take arbitrary differential operators.

yes that's great, and indeed you are moving the problem of computing these objects inside the operator itself, not in McState. The analogy here is that find_conn is defined in the operator not in McState, so I would do something similar also in continous space

**Comment by gcarleo** [comment]
anyways, I leave this to you, if you want to have something that works and then think about the interface of operators later on, that's completely understandable as well 

I would just suggest to rename `ContinuousEnergyExpectation` to something that does not make reference to the Expectation, but only to the operator, so that in the future we don't need to change names if we adopt a nicer general interface for operators 

`KineticPotential` might be a possible name `KineticDiagonal`, `KineticInteraction` etc are other options

File: netket/operator/_continuous_operator.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        basis = jnp.eye(x_in.shape[0])
+
+        dp_dx2 = jnp.trace(jax.vmap(hvp, in_axes=(None, 0, None), out_axes=0)(dlogpsi_x, basis, x_in))
+        """
+        dp_dx = dlogpsi_x(x_in) ** 2
+
+        return -0.5 * jnp.sum(dp_dx2 + dp_dx)
+
+    def totenergy(self, log_psi, params, x_in):
```
Comment (gcarleo) [inline netket/operator/_continuous_operator.py:62]: total_energy

File: netket/vqs/mc_expect_grad.py [hunk 8]
```diff
@@ -373,3 +393,58 @@ def gradfun(der_loc_vals, der_logs_ave):
...
+        out_axes=0,
+    )
+    x = x.reshape((-1, 10, x_shape[-1]))
+
+    def compute_kernel(i, x):
+        Oloc = local_value_vmap(parameters, x)
+
+        return i, Oloc
+
+    _, O_loc = jax.lax.scan(compute_kernel, 0, x)
```
Comment (PhilipVinc) [inline netket/vqs/mc_expect_grad.py:428]: why are you using a scan here?

**Comment by PhilipVinc** [comment]
> ink this should work following the same dispatch path, otherwise we need to introduce a different dispatch path for all possible hamiltonians in continous space, which is not very nice, I think

The same dispatch rules are used for all subtypes of a certain type. So it's only about defining the right types.
But we cannot yet treat hamiltonians as sum of different sub-terms (otherwise the variance is off) so we cannot rely on this. The full continuous hamiltonian must be a single operator. So it cannot reuse machinery from the discrete case.

> I mean, I understand why expect_and_grad works using dispatch in cases where you have density matrices or other objects that require specific expressions, but if we are doing expect and grads of the same family of (hermitean) operators both on the continuum and on the lattice

half the motivation of implementing dispatch there was to support continuous operators (and weird objects like Covariances and such).
While potential terms are trivial to implment with an interface like `get_conn`, the kinetic term is not. You need special code to handle it and to handle it with an MCState. And it would probably change when working with a density operator. 
You can have `if/else` statements checking for the type, or you can use dispatch. It's essentially the same solution (though dispatch is cleaner and easier to extend).

Example: in the current code the continuous operator has a function `def e_kin(self, log_psi, params, x_in):`. 
While ok for prototyping, this is conceptually wrong for the following reasons, in my opinion: 
 - `netket.operator` does not know nothing about MCState and flax/jax/whatever. This function introduces a dependency on the other object.
 - This `e_kin` will only work with pure states. It won-t work for mixed states. Or different kind of states we might invent in the future (such an ExactPureState or something else).
 - If you don't want to use dispatch, the only solution is to use `if isinstance(.., MCState)` or `if isinstance(ContinuousOperator)`... which is reinventing the wheel of dispatch.

> The analogy here is that find_conn is defined in the operator not in McState, so I would do something similar also in continous space

multiple dispatch is about having the rule live in the intersection between the `operator` *and* `MCState`.

--

In my opinion, to keep this clean, the ideal solution would be to remove those methods which are too specific and are not user-facing from `ContinuousOperator` (we need a better name, of course).

Surely they should at least be internal (start with a `_`).

Though I would rather make them free floating functions. they can stay defined in the same file.
Then a file `netket/vqs/mcstate_continouus.py` can import them and define those dispatch rules.

I would like to avoid things like `    kernel = Ô.totenergy` because that's exactly what we need to avoid (this cannot discriminate between needed for density matrices and pure states). Instead the dispatch rule for `def expect_and_grad(state: MCState, operator: ContinuousOperator...)` should just use the right kernel. in that way in the future it is possible to define 
```
def expect_and_grad(state: MCMixedState, operator: ContinuousOperator...):
    kernel = the_correct_mixed_kjernel

```

File: netket/vqs/mc_expect_grad.py (hunk 8 above)
Comment (gpescia) [reply netket/vqs/mc_expect_grad.py:428]: computing the kinetic energy needs  a lot of memory because of the second order derivative. So this was a try to reduce memory consumption. Instead of feeding all samples at once I feed each sample individually (or in this case 10 samples a time to be precise)

**Comment by gcarleo** [comment]
> If you don't want to use dispatch, the only solution is to use if isinstance(.., MCState) or if isinstance(ContinuousOperator)... which is reinventing the wheel of dispatch.

> netket.operator does not know nothing about MCState and flax/jax/whatever. This function introduces a dependency on the other object.

The conceptual issue here is that, contrary to the discrete case, the operator in a sense **depends** on the quantum state, because it is a differential operator acting on functions, so introducing some dependency on JAX/Flax in the operator itself seems inevitable to me. The elegant alternative would be to define a lazy differential operator (again, these operators are **not** sparse matrices) that is abstract and does not return JAX/Flax objects. (we could do that, but it increases complexity/ how can we do it nicely?) 

That's why I was suggesting a simpler alternative which is to ask the general interface of operators to be able to implement D(X)=log(<X|A|Psi>) (dispatching possibly on Psi, that might be a pure or a mixed state). 
If you do that, expect_and_grad can use the same path for both continuum and lattice, because the local estimators are easily written in terms of D(X), without any reference to whether X is discrete or continous.  

The other advantage is that if you change hamiltonians in the continuum (say, you have another differential operator or you have a relativistic hamiltonian or something...) you just define a new operator and its D(X). 

My impression is that instead in the current implementation if you want to do that, you would need to add a further dispatch rule in MCState, and it would not be clear from the definition itself of the operator what the differential operator is. In other words, you would need to look into MCState (why?) to know if one is using a second derivative or a first derivative etc. 

Also, D(X) is more general than expect and grad, and could be used when working with operators independently than for computing expectation values.

**Comment by gcarleo** [comment]
A possibility is to make the continous operator return directly a JAX differential operator, such as grad or something, that later can be used in MCState. The issue was that if you just return the laplacian, then somehow in expect and grad you have to make this differential operator act on log psi and not psi itself (whereas the natural definition of the differential operator is on psi, not log psi)

**Comment by PhilipVinc** [comment]
> That's why I was suggesting a simpler alternative which is to ask the general interface of operators to be able to implement D(X)=log(<X|A|Psi>) (dispatching possibly on Psi, that might be a pure or a mixed state).

Yes, that's the way forward.
But I think that, to implement this correctly, we'd need to sit down and define carefully the interface and how the differential objects behave. Bcause they shuld also work with discrete operators.
It will take quite a bit of experimentation, and it can be done together with the rewrite of the operators.

> The other advantage is that if you change hamiltonians in the continuum (say, you have another differential operator or you have a relativistic hamiltonian or something...) you just define a new operator and its D(X).

Nothing prevents you from doing something similar with the current interface, that is, just write the new operator and its kernel function.
This is what the various (discrete/Squared/Transposed) operators do at this point.
To make this work on continuous operators we just need an abstract class `AbstractContinuousOperator` and to define the dispatch rules on it instead of the specific `ContinuousKineticHamiltonian`.

> My impression is that instead in the current implementation if you want to do that, you would need to add a further dispatch rule in MCState, and it would not be clear from the definition itself of the operator what the differential operator is

We need a type hierarchy. Then you know what the differential operators are. 

--

I really think that for this PR the minimum changes should be implemented to make continuous stuff work.

We should discuss about how to define the interface for differentials
Just not here.
Let's open another Issue about it and sketch the implementation?

**Comment by PhilipVinc** [comment]
my god.
I think something gone wrong

File: netket/models/DeepSets.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
+from typing import Union, Tuple, Any
```
Comment (gcarleo) [inline netket/models/DeepSets.py:1]: I would not include the DeepSet for now, since it is still under investigation/ we don't have a standard model for this yet

File: netket/operator/_continuous_diffoperator.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+from netket.utils.types import DType
+
+from netket.hilbert import AbstractHilbert
+from netket.operator import AbstractOperator
+
+import jax
+import jax.numpy as jnp
+
+
+class DiffOperator(AbstractOperator):
```
Comment (gcarleo) [inline netket/operator/_continuous_diffoperator.py:11]: I would prefer a more explicit name, say DifferentialOperator (Diff is a bit ambiguous ?)

Comment (gcarleo) [reply netket/operator/_continuous_diffoperator.py:11]: Actually, where is this DiffOperator used?

Comment (gpescia) [reply netket/operator/_continuous_diffoperator.py:11]: it is not used yet. but it is the more general interface which could replace the KineticPotential if the implementation is all right with you...

Comment (PhilipVinc) [reply netket/operator/_continuous_diffoperator.py:11]: I think we should sit down a moment and define the API of this interface on monday/tuesday.

Like, how it should behave under composition/addition/ecc so that we can define only one `expect/grad` rule that works for all continuous operators...

Comment (PhilipVinc) [reply netket/operator/_continuous_diffoperator.py:11]: I really like the idea, but I think that like this it is too crude?

File: netket/models/DeepSets.py (hunk 9 above)
Comment (PhilipVinc) [reply netket/models/DeepSets.py:1]: I do agree with @gcarleo but if we merge this can we have at least a toy model that works for a toy problem?

Comment (PhilipVinc) [reply netket/models/DeepSets.py:1]: So that it is a bit more clear for whoever decides to play with this what the model should do.

Comment (gcarleo) [reply netket/models/DeepSets.py:1]: yes I would just do the harmonic oscillator with a gaussian ansatz in the notebook, no deepset involved...

**Comment by PhilipVinc** [comment]
why was this closed?
